src/eos-updater-poll-common.c

```diff
@@ -0,0 +1,910 @@
+    {
+      message ("Product-specific branch configuration found");
+      if (!process_single_group (bkf, group_name, out_on_hold, out_ref, error))
+        return FALSE;
+      if (*out_on_hold)
+        message ("Product is on hold, nothing to upgrade here");
+      return TRUE;
+    }
+  /* Check for a DEFAULT_GROUP entry */
+  if (g_key_file_has_group (bkf, DEFAULT_GROUP))
+    {
+      message ("No product-specific branch configuration found, following %s",
+               DEFAULT_GROUP);
+      if (!process_single_group (bkf, DEFAULT_GROUP, out_on_hold, out_ref, error))
+        return FALSE;
+      if (*out_on_hold)
+        message ("No product-specific configuration and %s is on hold, "
+                 "nothing to upgrade here", DEFAULT_GROUP);
+      return TRUE;
+    }
+
+  *out_on_hold = FALSE;
+  *out_ref = NULL;
+  return TRUE;
+}
+
+gboolean
+get_upgrade_info_from_branch_file (EosBranchFile *branch_file,
+                                   gchar **upgrade_refspec,
+                                   gchar **original_refspec,
+                                   EosMetricsInfo **metrics,
+                                   GError **error)
+{
+  g_autoptr(OstreeDeployment) booted_deployment = NULL;
+  g_autofree gchar *booted_refspec = NULL;
+  g_autofree gchar *booted_remote = NULL;
+  g_autofree gchar *booted_ref = NULL;
+  g_autoptr(GHashTable) hw_descriptors = NULL;
+  g_autofree gchar *vendor = NULL;
+  g_autofree gchar *product = NULL;
+  g_autofree gchar *product_group = NULL;
+  g_autofree gchar *upgrade_ref = NULL;
+  gboolean on_hold = FALSE;
+
+  g_return_val_if_fail (EOS_IS_BRANCH_FILE (branch_file), FALSE);
+  g_return_val_if_fail (upgrade_refspec != NULL, FALSE);
+  g_return_val_if_fail (original_refspec != NULL, FALSE);
+
+  booted_deployment = eos_updater_get_booted_deployment (error);
+
+  if (!get_origin_refspec (booted_deployment, &booted_refspec, error))
+    return FALSE;
+
+  if (!ostree_parse_refspec (booted_refspec, &booted_remote, &booted_ref, error))
+    return FALSE;
+
+  hw_descriptors = get_hw_descriptors ();
+  vendor = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, VENDOR_KEY)));
+  product = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, PRODUCT_KEY)));
+
+  product_group = g_strdup_printf ("%s %s", vendor, product);
+  if (!process_branch_file (branch_file->branch_file, product_group, &on_hold, &upgrade_ref, error))
+    return FALSE;
+
+  if (on_hold)
+    upgrade_ref = g_strdup (booted_ref);
+  else
+    {
+      if (upgrade_ref == NULL)
+        {
+          message ("No product-specific branch configuration or %s found, "
+                   "following the origin file", DEFAULT_GROUP);
+          upgrade_ref = g_strdup (booted_ref);
+        }
+
+      message ("Using product branch %s", upgrade_ref);
+      *upgrade_refspec = g_strdup_printf ("%s:%s", booted_remote, upgrade_ref);
+      *original_refspec = g_strdup (booted_refspec);
```

Comment:
Snippet from the original function - https://github.com/endlessm/eos-updater/blob/master/src/eos-updater-util.c#L469

From eos-updater-util, of course.